crates/build/src/docker.rs

```diff
@@ -0,0 +1,563 @@
+    let mut cmds = vec![format!(
+        "mkdir -p target && cargo contract build {} --output-json > {}",
+        build_args, b_path_str
+    )];
+
+    entrypoint.append(&mut cmds);
+
+    // in order to optimise the container usage
+    // we are hashing the inputted command
+    // in order to reuse the container for the same permutation of arguments
+    let mut s = DefaultHasher::new();
+    entrypoint.hash(&mut s);
+    let digest = s.finish();
+    // taking the first 5 digits to be a unique identifier
+    let digest_code: String = digest.to_string().chars().take(5).collect();
+
+    let container_name = format!("ink-verified-{}-{}", contract_name, digest_code);
+
+    let mut filters = HashMap::new();
+    filters.insert("name".to_string(), vec![container_name.clone()]);
+
+    let containers = client
+        .list_containers(Some(ListContainersOptions::<String> {
+            all: true,
+            filters,
+            ..Default::default()
+        }))
+        .await?;
+
+    let container_option = containers.first();
+
+    let mount = Mount {
+        target: Some(String::from(MOUNT_DIR)),
+        source: Some(
+            host_folder
+                .to_str()
+                .context("Cannot convert path to string.")?
+                .to_string(),
+        ),
+        typ: Some(MountTypeEnum::BIND),
+        ..Default::default()
+    };
+    let host_cfg = Some(HostConfig {
+        mounts: Some(vec![mount]),
+        ..Default::default()
+    });
+
+    let user;
+    #[cfg(unix)]
+    {
+        user = Some(format!(
+            "{}:{}",
+            users::get_current_uid(),
+            users::get_current_gid()
+        ));
+    };
+    #[cfg(windows)]
+    {
+        user = None;
+    }
+
+    let mut labels = HashMap::new();
+    labels.insert("cmd_digest".to_string(), digest.to_string());
+    let config = Config {
+        image: Some(build_image.id.clone()),
+        entrypoint: Some(entrypoint),
+        cmd: None,
+        labels: Some(labels),
+        host_config: host_cfg,
+        attach_stderr: Some(true),
+        tty: Some(true),
+        user,
+        ..Default::default()
+    };
+    let options = Some(CreateContainerOptions {
+        name: container_name.as_str(),
+        platform: Some("linux/amd64"),
+    });
+
+    let container_id = match container_option {
+        Some(container) => {
+            container
+                .id
+                .clone()
+                .context("Container does not have an ID")?
+        }
+        None => client.create_container(options, config).await?.id,
+    };
+
+    client
+        .start_container::<String>(&container_id, None)
+        .await?;
+
+    maybe_println!(
+        verbosity,
+        " {} {}\n {}",
+        format!("{build_steps}").bold(),
+        "Started the build inside the container"
+            .bright_green()
+            .bold(),
+        "You can close this terminal session. The execution will be finished in the background"
```

Comment:
I find this confusing when it prints this, even though it is true. Why would we tell a user to do this, especially when we are still waiting for a success or failure of the build?

In my mind we should be attaching to the docker run anyway, so we can see any errors produced by the build. e.g. if there is a compilation error in the contract we just get:

![image](https://github.com/paritytech/cargo-contract/assets/75586/204a08ee-1c2d-48da-bca0-a984ddec9980)

Reply:
Yes, I will attach the error message if there is any. However, since the docker container runs the build with the `--output-json` there are any build logs produced other than errors

Reply:
```
I find this confusing when it prints this, even though it is true. Why would we tell a user to do this, especially when we are still waiting for a success or failure of the build?
```
Because this is the behaviour, the container will continue to run if the user quits the session with `ctrl + c`. So, I am just giving the user information.

Reply:
I, as a user, found it confusing. Even though as you say it is technically correct.

Also, if the user were to close the terminal session as suggested, then the build result would not be rewritten and the resulting `.bundle` file would be incorrect?

Anyway, let's attach to the docker process and keep it as similar as possible to a regular build.

Reply:
Actually, the bundle will be correct since it doesn't contain any paths. All the build result shows is the target dirs of the artefacts and the code size. If we exit terminal early, the build result will just not be printed in the terminal. This is equivalent of just running the build in the background

Reply:
As I read the, once the build has succeeded it rewrites some paths and updates the image used in the metadata https://github.com/paritytech/cargo-contract/blob/5ec303bdc62e652dd84be3d4475d6b97db2781b7/crates/build/src/docker.rs#L162C29-L162C29.

So if you quit the terminal it will never do that step. True that the contract files will still be there but isn't the main reason one runs `--verifiable` to be able to compare the output of builds, and to know which image was used.

Reply:
All we can do in this case is not to display the "you can close the terminal" message

Reply:
Let's remove the message then :+1: